Amazon SQS 5.9 (#1192)
mulesoft/docs-connectors: amazon-sqs/5.9/modules/ROOT/pages/amazon-sqs-connector-config-topics.adoc

= Amazon SQS Connector 5.9 Additional Configuration Information - Mule 4
:page-aliases: connectors::amazon/amazon-sqs-connector-config-topics.adoc

> 

Reviewer: I would add an intro sentence into this doc, something like:

_Set additional configuration for Anypoint Connector for Amazon SQS (Amazon SQS Connector), such as controlling polling behavior globally and locally, receiving messages on the Primary Node Only, creating the FIFO queue, and managing large payloads_

> == Long and Short Polling

Reviewer: ```suggestion
== Configure Polling Behavior
```

Since version 5.9, the default behavior of the Receive Messages source and the Read operation is to use long polling:

* With long polling, SQS Connector returns the query results as soon as possible, if any messages are present.
+
Otherwise, SQS Connector waits the specified amount of seconds before returning an empty response.
* With short polling, SQS Connector always returns the query results as soon as possible, even if that means returning an empty response.

Long polling can reduce the number of network requests without affecting the throughput.

> You can control polling behavior either globally or at the individual source and operation level.

Reviewer: ```suggestion
You can configure SQS Connector to control polling behavior either globally or locally, and at the operation level.
```

Reviewer: Though, at the operation level is not explained later in the next section. Maybe you can remove it, because the explanation mentions only the Global Configuration Element of Studio.

> === Control Polling Behavior Globally

Reviewer: ```suggestion
=== Configure Control Polling Behavior Globally
```

> To control polling behavior globally, use the Default Wait Time field on the Advanced tab of the global element configuration screen for the Receive Messages source, Read operation, or both.
> 
> image::amazon-sqs-config-wait-time.png[Default Wait Time Configuration]
> 
> Set this field to an integer value between `1` and `20` (the default) to specify long polling as the default behavior. The integer specifies how long to wait, in time units specified in the Time unit field, before returning an empty response.
> 
> Set this field to `0` to specify short polling as the default behavior.

Reviewer: ```suggestion
To configure SQS Connector to control polling behavior globally, follow the next steps:

. In the Studio flow of your application, select either the *Receive Messages* source or the *Read* operation.
. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of the source or operation in the app.
. In the *Advanced* tab, set the *Default Wait Time* field to an integer value between `1` and `20` (the default) to specify long polling as the default behavior, or set this field to `0` to specify short polling as the default behavior.
+
The integer value specifies how long to wait, in time units specified in the *Time unit* field, before returning an empty response.
+
image::amazon-sqs-config-wait-time.png[Default Wait Time Configuration]

```

Author: I added more details, but not as steps.

> === Control Polling behavior Locally

Reviewer: ```suggestion
=== Configure Control Polling Behavior Locally
```

To control polling behavior locally, use the Wait Time field on the Advanced tab of the Receive Messages source properties window, the Read operation properties window, or both.

> *Question for Reviewers asked in the Slack channel. Why are the polling fields missing for the Read operation in Studio? Only the Wait Time field is showing up there.*

Reviewer: Based on the the Wait Time description of the reference:
_Set this field to a value between `1` and `20` to use long polling. With long polling, if there are messages available immediately, the connector returns the query results immediately. If not, the connector waits up to the specified amount of seconds for a message to be added before returning an empty response.
Set this field to `0` to use short polling. With short polling, the connector always returns the query results as soon as possible, even if that means returning an empty response._

You could write something similar as to what I mentioned before for Default Wait Time

[node-behavior]

> == Receive Messages on the Primary Node Only

Reviewer: ```suggestion
== Configure How to Receive Messages on the Primary Node Only
```

Author: I don't think this is necessary.

The default behavior of the Receive Messages source is to receive messages on all nodes when running in cluster mode.

If you want the app to receive messages on the primary node only:

. In Studio, select the *Receive Messages* source for your flow.
. On the *Advanced* sub-tab of the *Receive messages* tab, select the *Primary node only* checkbox:
+
image::amazon-sqs-primary-node-only.png[Read message configuration Advanced tab with Primary node only option selected]

The multi-node cluster behavior depends on the selected SQS queue type. When using the Standard queue, SQS guarantees that a message is delivered at least once. However, duplicate messages might be introduced in the queue. To avoid duplication of messages in the queue, you can use a FIFO (First-In-First-Out) queue to ensure that a message is delivered only once and that it remains available until a consumer processes and deletes it.

For more information about queue types, see https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/standard-queues.html[AWS SQS Stanard Queues] and https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/FIFO-queues.html[AWS SQS FIFO Queues].

> == Create the FIFO Queue

Reviewer: ```suggestion
== Configure FIFO Queues
```

FIFO queues deliver messages in the order in which they are sent.

> To create the FIFO queue, use the *Create queue* operation in the connector, and set the *FifoQueue* and *ContentBasedDeduplication* field values to true.

Reviewer: ```suggestion
To create the FIFO queue, use the *Create queue* operation in the connector, and set the *FifoQueue* and *ContentBasedDeduplication* field values to `true`.
```

End the *queue name* value with the `.fifo` extension, for example: +
`MyTestFIFOQueue.fifo`.

When you create the FIFO queues, select any of the 15 regions that Amazon supports.

When you use Amazon SQS Connector to send messages to the FIFO queue, use the same queue URL and region name that you used when creating the FIFO queue. You must also provide a value for the attribute `message group id` in the connector configuration while sending messages.

> == Managing Large Payloads

Reviewer: ```suggestion
== Configure Managing Large Payloads
```

Author: Changed to Manage Large Payloads
